src/app/src/reducers/FiltersReducer.js

```diff
@@ -7,15 +7,109 @@ import {
     updateContributorTypeFilter,
     updateCountryFilter,
     resetAllFilters,
+    updateAllFilters,
 } from '../actions/filters';
 
+import {
+    completeFetchContributorOptions,
+    completeFetchContributorTypeOptions,
+    completeFetchCountryOptions,
+} from '../actions/filterOptions';
+
 const initialState = Object.freeze({
     facilityName: '',
     contributors: Object.freeze([]),
     contributorTypes: Object.freeze([]),
     countries: Object.freeze([]),
 });
 
+const maybeSetContributorOptionsFromQueryString = (state, payload) => {
+    if (!state.contributors.length) {
+        return state;
+    }
+
+    // filter out any options set from the querystring which turn out
+    // not to be valid according to the API's response
+    const updatedContributors = state
+        .contributors
+        .reduce((accumulator, { value }) => {
+            const validOption = payload
+                .find(({ value: otherValue }) => value === otherValue);
+
+            if (!validOption) {
+                return accumulator;
+            }
+
+            return accumulator
+                .concat(Object.freeze({
+                    value,
+                    label: validOption.label,
+                }));
+        }, []);
+
+    return update(state, {
+        contributors: { $set: updatedContributors },
+    });
+};
+
+const maybeSetContributorTypeOptionsFromQueryString = (state, payload) => {
+    if (!state.contributorTypes.length) {
+        return state;
+    }
+
+    // filter out any options set from the querystring which turn out
+    // not to be valid according to the API's response
+    const updatedContributorTypes = state
+        .contributorTypes
+        .reduce((accumulator, { value }) => {
+            const validOption = payload
```

Comment:
The implementation of these reduce functions is nearly the same, but each of them needs access to the distinct `payload` array, so it wasn't possible to extract a function like `(acc, { value }) => acc.concat(value);` or similar to share across them.